dev/benchmarking/README.md

```diff
@@ -0,0 +1,305 @@
+  (The `--watch` flag on the command line. Not recommended for production.)
+- An endpoint that receives POST requests only (localhost:5000/graphql).
+
+#### Ingestion
+
+The script [graphql_submit.sh](./graphql_submit.sh) simplifies submission
+to the database via curl. Examples:
+
+```shell
+./graphql_submit.sh benchmarks examples/benchmark_example.json 
+./graphql_submit.sh results examples/benchmark_result_example.json
+```
+
+#### Querying
+
+The output of the query is a JSON object that is hard to read on the command line.
+Here is an example query in the shell:
+```shell
+curl -X POST \
+  -H "Content-Type: application/json"  \
+  --data '{"query": "{projectDetails{ projectName }}"}' \
+  localhost:5000/graphql
+```
+which yields
+```
+{"data":{"projectDetails":{"projectName":"Apache Arrow"}}}
+```
+
+Here is an example query using Python:
+```python
+>>> import json
+>>> import requests
+>>>
+>>> uri = "http://localhost:5000/graphql"
+>>> query = json.load(open("examples/graphql_query_environment_view.json"))
+>>> result = requests.post(uri, json=query)
+>>> message = "{benchmarkLanguage} {languageVersion} {dependencies}"
+>>>
+>>> for row in result.json()['data']['allEnvironmentViews']['edges']:
+...     print(message.format(**row['node']))
+...
+Python 2.7 {"six":"","numpy":"1.14","other_lib":"1.0"}
+Python 2.7 {"six":"","numpy":"1.15","other_lib":"1.0"}
+Python 3.6 {"boost":"1.42","numpy":"1.15"}
+```
+
+## Deployment
+
+(work in progress).
+
+
+## Quick reference
+
+- String variables `'have single quotes'`
+- Arrays `'{"have", "curly", "braces"}'::text[]` or `'{1, 2, 3}'::integer[]`
+- JSONb `'{"has":"this", "format":42}'::jsonb`
+- When comparing nullable values use `x IS DISTINCT FROM y` rather than `x = y`
+
+
+## Data model
+
+The primary view in the database is the
+[`benchmark_result_view`](#benchmark_result_view),
+which joins the actual benchmark result data with its context:
+
+- [`machine_view`](#machine_view): Hardware and OS
+- [`evironment_view`](#environment_view): Build environment
+  (language, dependencies)
+- [`benchmark_view`](#benchmark_view): Benchmark definition
+  (type and units, e.g. 'Time', 'ms')
+
+The following views can be inserted to directly: trigger
+functions (in ['5_triggers.sql'](ddl/5_triggers.sql))
+will look up the relevant ID numbers machine ID, benchmark ID and other
+ID numbers to use when inserting a new result.
+
+
+### machine_view
+
+The machine  environment (CPU, GPU, OS) used for a reported benchmark result.
+- "machine_name" is unique in the "machine" table
+- "gpu_model_name" is unique in the "gpu" (graphics processing unit) table
+    - Empty string (''), not null, is used for machines that won't use the GPU
+- "cpu_model_name" is unique in the "cpu" (central processing unit) table
+- "os_name", "os_architecture_name", and "os_kernel_name"
+  are unique in the "os" (operating system) table
+
+Column                  |  Type   | Nullable | Note
```

Comment:
are these autogenerated somehow?  If not I'd be afraid they would go out of date with the actual schema.

Reply:
They are not! Your concern is valid. I have replaced the 'Data model' section with a hyperlink to a separate (to-be) autogenerated file 'DataModel.md' (that currently still contains manual entries) until I can get something working that does this...hopefully built along with the visualization you suggest.